New rule: role is permitted (#2084)
auto-wcag/auto-wcag: _rules/aria-role-permitted-j7zzqr.md

---
id: j7zzqr

> name: ARIA role is permitted.

Reviewer: ```suggestion
name: ARIA role is permitted for the element
```

rule_type: atomic
description: |
  This rule checks that WAI-ARIA roles are allowed for the element they are specified on.
accessibility_requirements:
  html-aria:docconformance:
    title: ARIA in HTML, 4. Document conformance requirements for use of ARIA attributes in HTML
    forConformance: true
    failed: not satisfied
    passed: satisfied
    inapplicable: satisfied
  wcag20:1.3.1: # Info and Relationships (A)
    secondary: true
  wcag20:4.1.2: # Name, Role, Value (A)
    secondary: true
  wcag-technique:ARIA4: # Using a WAI-ARIA role to expose the role of a user interface component

>     secondary: true

Reviewer: I'm not sure these should be here. I think the correlation between this rule and WCAG is pretty weak. The one failed example this rule has, I think anyone who would fail that under 1.3.1 and 4.1.2 would be wrong to do so.

There might be cases I didn't think of, but while we didn't constrain secondary requirements to only be used when there are corresponding examples, that is a fairly strong indicator, and perhaps even something we'd want to do in our authoring policy; state that if there is a secondary requirement, there should be examples that demonstrate it.

@kengdoj what's your thoughts about using secondary here?

input_aspects:
  - Accessibility Tree
  - DOM Tree
acknowledgments:
  authors:
    - Jean-Yves Moyen
---

## Applicability

This rule applies to any [HTML element][namespaced element] that is [included in the accessibility tree][] and has an [explicit semantic role][explicit role].

## Expectation

> For each test target, its [explicit semantic role][explicit role] is allowed on this element, according to [ARIA in HTML specifications][aria in html document conformance].

Reviewer 2: ```suggestion
For each test target, its [explicit semantic role][explicit role] is allowed on this element, according to the [ARIA in HTML specification][aria in html document conformance].
```

## Assumptions

There are no assumptions.

## Accessibility Support

There are no accessibility support issues known.

## Background

> The presence of an invalid role often implies that the programmatic role do not correspond to the one that is conveyed visually, or that the interactions provided by the elements do not match the ones expected for this role. Therefore, both [Success Criterion 1.3.1 Info and Relationships][sc131] and [Success Criterion 4.1.2 Name, Role, Value][sc412] are secondary requirements for this rule.

Reviewer: I'm not sure how true this really is. See my comment above. Also, typo:

```suggestion
The presence of an invalid role often implies that the programmatic role does not correspond to the one that is conveyed visually, or that the interactions provided by the elements do not match the ones expected for this role. Therefore, both [Success Criterion 1.3.1 Info and Relationships][sc131] and [Success Criterion 4.1.2 Name, Role, Value][sc412] are secondary requirements for this rule.
```

We probably also want to say something that this rule is stricter than 1.3.1 and 4.1.2, if we keep those as secondary.

### Related rules

- [Role attribute has valid value](https://www.w3.org/WAI/standards-guidelines/act/rules/674b10/proposed/) checks that the value of the `role` attribute exists in ARIA, while this rule checks that it is allowed on the element using it.

### Bibliography

- [Document conformance requirements for use of ARIA attributes in HTML](https://www.w3.org/TR/html-aria/#docconformance)
- [Understanding SC 1.3.1: Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html)
- [Understanding Success Criterion 4.1.2: Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value.html)
- [Technique ARIA4: Using a WAI-ARIA role to expose the role of a user interface component](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA4)

## Test Cases

### Passed

#### Passed Example 1

This `a` element has an [explicit role][] of `button`, which is allowed.

```html
<a href="https://www.w3.org/WAI/standards-guidelines/act/rules/" role="button">All ACT rules</a>
```

#### Passed Example 2

This `h1` element has an [explicit role][] of `tab`, which is allowed

```html
<h1 role="tab">ACT rules</h1>
```

> ### Failed

Reviewer: I think I'd like a few more examples here of different types of elements. Maybe have a landmark and a heading?

---
id: j7zzqr
name: ARIA role is permitted for the element.
rule_type: atomic
description: |
  This rule checks that WAI-ARIA roles are allowed for the element they are specified on.
accessibility_requirements:
  html-aria:docconformance:
    title: ARIA in HTML, 4. Document conformance requirements for use of ARIA attributes in HTML
    forConformance: true
    failed: not satisfied
    passed: satisfied
    inapplicable: satisfied
input_aspects:
  - Accessibility Tree
  - DOM Tree
acknowledgments:
  authors:
    - Jean-Yves Moyen
---

## Applicability

This rule applies to any [HTML element][namespaced element] which has an [explicit semantic role][explicit role].

## Expectation

For each test target, its [explicit semantic role][explicit role] is allowed on this element, according to the [ARIA in HTML specification][aria in html document conformance].

## Assumptions

There are no assumptions.

## Accessibility Support

There are no accessibility support issues known.

## Background

> [ARIA in HTML][aria in html document conformance] also defines the [implicit semantic role][implicit role] of each element. Setting the [explicit role][] as the same as the [implicit one][implicit role] is not recommended but nonetheless allowed. This rule doesn't use that in any of its test cases.

Reviewer 3: ```suggestion
[ARIA in HTML][aria in html document conformance] also defines the [implicit semantic role][implicit role] of each element. Setting the [explicit role][] as the same as the [implicit role][] is not recommended but nonetheless allowed. This rule does not apply in that scenario.
```

Reviewer 4: aria in html does not define the implicit roles.  HTML AAM does.  ARIA in HTML informatively identifies what these are.

> This rules apply to every element, even if they are not [included in the accessibility tree][]. This is because the roles of `none` or `presentation` are only allowed on certain elements. If the rule was only looking at content [included in the accessibility tree][], it wouldn't flag incorrect use of these roles. 

Reviewer 3: If an element and its children are not included in the accessibility tree, it doesn't really matter what the roles are. Therefore, it shouldn't be an accessibility problem if these roles are incorrectly used. While I agree that it's technically incorrect, I think it should be considered more as a recommendation rather than a violation - similar to the use of an explicit role that is redundant with the implicit role.

@scottaohara Care to comment on the ARIA in HTML position on this one?

Reviewer 4: i mean, yes - if the element is hidden form the accessibility tree then sure, no one can access it / there can't be an issue.

the counter point being that if it's rendered in the DOM but marked as hidden, is that because at some point it would be expected that a user could then unhide that content to interact with it?  Could excluding something just because it's temporarily hidden not cause a bunch of gotchas for developers who might run a scan, it comes up clean, and then they still get bugs because someone had the audacity to try and use their UI? :)

but seriously, i can see the rational to just expose this as a recommendation / needs review - in the hidden state, as long as if it becomes shown it then would be marked as a violation.  it just seems like a delay in flagging some issues because maybe someone did a hack-job at 'removing' an errant element, and _that_ shouldn't be called out?

Reviewer 3: @scottaohara With the way that UIs are built these days, what is in the DOM changes with almost every state change. There are often scaffolds that are hidden, and then made whole when they're actually used / made visible. We don't report on a lot of issues in hidden content because of this - teams were frustrated by all of the false positives. 

While I agree that it's unlikely that an incorrectly used role will be corrected on such a state change, I don't think we should call out a tool as incorrect for ignoring that content completely.

```

#### Failed Example 2

This `aside` element has an [explicit role][] of `navigation`, which is not allowed.

```html
<aside role="navigation">
  <a href="https://www.w3.org">W3C</a> <a href="https://www.w3.org/TR/WCAG21/">WCAG 2.1</a> <a href="https://www.w3.org/WAI/standards-guidelines/act/rules/"> ACT rules</a>
</aside>
```

#### Failed Example 3 

These `h1` elements have an [explicit role][] of `listitem`, which is not allowed; the `div` element has an [explicit role][] of `list`, which is allowed.

```html
Fruits:
<div role="list">
  <h1 role="listitem">Apple</h1>
  <h1 role="listitem">Banana</h1>
  <h1 role="listitem">Orange</h1>
</div>
```

#### Failed Example 4

The first `li` element has an [explicit role][] of `presentation`, which is not allowed.

```html
Fruits:
<ul>
  <li role="presentation">Apple</li>
  <li>Banana</li>
  <li>Orange</li>
</ul>
```

### Inapplicable

#### Inapplicable Example 1

There is no [HTML element][namespaced element].

```svg
<svg xmlns="http://www.w3.org/2000/svg">
  <title>This is an SVG</title>
</svg>
```

#### Inapplicable Example 2

This `button` element is not [included in the accessibility tree][].

```html

> <button role="list" style="display:none;">Click me</button>

Reviewer 5: From the background:

> This rules apply to every element, even if they are not [included in the accessibility tree][].

Wouldn't that make this a failed example instead of an inapplicable example?

Reviewer 6: I agree with @tbostic32 

The applicability are elements with an explicit role defined which is the case.

Reviewer 3: See my other comment. I do think this should be inapplicable to this rule, and the note and applicability rather than this testcase should be modified. However, one or the other needs to change.

Author: True, moved it to an Failed case.

I think that this is still failing ARIA in HTML, even if it doesn't cause an accessibility problem in the end, so I'd rather keep it that way. ARIA in HTML doesn't specify that the roles are only allowed on elements that are exposed. Moreover, if we restrict the rule to exposed elements, suddenly Failed Example 4 (`<li role="presentation">`) becomes Inapplicable when it's a clear failure since the only allowed role is `listitem`.

This is a bit similar to the fact that, in HTML¹, duplicate `id` are not allowed even though they might very well never cause a single problem. It is nonetheless invalid HTML. Here also, as far as I understand, using the wrong role on a hidden element is invalid ARIA, even if the user never encounter that.

¹: this is not about SC 4.1.1, but HTML/DOM specs

Reviewer 7: In accordance with my comment https://github.com/act-rules/act-rules.github.io/pull/2084#discussion_r1340254070, to me, Failed Example 4 is not failing because of the presence of the "presentation" role in a DOM children. Rather, its failure is attributed to the absence of an accessibility child with the role "listitem" within the DOM child.

Hence, it is not the "role=presentation" itself that impacts the test result; rather, the test fails due to the absence of a child with the "listitem" role.

I agree with Tom that we should consider elements included into the accessibility tree.

Author: @giacomo-petri The reason why `role="presentation"` was taken into account by this rule is that certain elements (e.g., `article` or `aside`) explicitly allow it in the list of allowed roles; while some other (like `button`) also have an explicit list of allowed roles but **without** `presentation` being listed.

On looking a bit more closely, it might be that the ones where `presentation` isn't listed are the elements which are normally focusable, so the conflict would trigger 🤔 I'll look a bit deeper.

Reviewer 4: just going to state one more time that i'd err on the side of flagging this - as otherwise it falls on making a developer/tester have to do more work to find errors, by ensuring they fully comb through a UI and reveal all hidden content.  I do acknoweldge there are probably instances of devs hiding problematic markup, rather than fixing it... but should a dev really be surprised to get an error flagged for something they knew was an error and 'hid' it as their 'fix'?

accessibility_requirements:
  html-aria:docconformance:
    title: ARIA in HTML, 4. Document conformance requirements for use of ARIA attributes in HTML
    forConformance: true
    failed: not satisfied
    passed: satisfied
    inapplicable: satisfied
input_aspects:
  - Accessibility Tree
  - DOM Tree
acknowledgments:
  authors:
    - Jean-Yves Moyen
---

## Applicability

This rule applies to any [HTML element][namespaced element] which has an [explicit semantic role][explicit role].

## Expectation

For each test target, its [explicit semantic role][explicit role] is allowed on this element, according to the [ARIA in HTML specification][aria in html document conformance].

## Assumptions

There are no assumptions.

## Accessibility Support

There are no accessibility support issues known.

## Background

Each element also has an [implicit semantic role][implicit role] defined in the [HTML Accessibility API Mappings][html aam:roles]. Setting the [explicit role][] as the same as the [implicit role][] is not recommended but nonetheless allowed. This rule does not check that scenario.

> This rule does not check specifically whether the [explicit role][] is deprecated, or whether it is the special `generic` role that should not be used by authors. These are not conformance requirement, or are checked by other rules.

Reviewer 4: ```suggestion
This rule does not check specifically whether the [explicit role][] is deprecated, or whether it is the special `generic` role that should not be used by authors. These are not conformance requirements, or are checked by other rules.
```

are deprecated features surfaced by other rule(s)?  If so, can there be a link to that?

> This rules apply to every element, even if they are not [included in the accessibility tree][]. This is because the roles of `none` or `presentation` are only allowed on certain elements. If the rule was only looking at content [included in the accessibility tree][], it wouldn't flag incorrect use of these roles.

Reviewer 4: ```suggestion
This rule applies to every element, even if they are not [included in the accessibility tree][]. This is because the roles of `none` or `presentation` are only allowed on certain elements. If the rule was only looking at content [included in the accessibility tree][], it wouldn't flag incorrect use of these roles.
```

this section should probably also mention content hidden from the a11y tree with aria-hidden=true, or other host language features (per the example on line 182)

Reviewer 7: I'm uncertain about the accuracy of the following statement:

> This rule applies to every element, even if they are not [included in the accessibility tree][]. This is because the roles of `none` or `presentation` are only allowed on certain elements. If the rule was only looking at content [included in the accessibility tree][], it wouldn't flag incorrect use of these roles.

According to section [7.3 Relationships in the Accessibility Tree](https://w3c.github.io/aria/#tree_relationships),

> The accessibility children of a DOM element are all of the children of that element's corresponding [accessible object](https://w3c.github.io/aria/#dfn-accessible-object) in the [accessibility tree](https://w3c.github.io/aria/#dfn-accessibility-tree). In terms of the DOM, that includes the following (with exclusions listed blow):
> - The DOM children of the [element](https://dom.spec.whatwg.org/#concept-element).
> - All DOM descendants of the [element](https://dom.spec.whatwg.org/#concept-element) with only elements of role [generic](https://w3c.github.io/aria/#generic) or [none](https://w3c.github.io/aria/#none) intervening.

This principle is equally applicable to the accessibility parent definition (https://w3c.github.io/aria/#dfn-accessibility-parent)

A clear illustration of this can be found in Example 43: [Example 43](https://w3c.github.io/aria/#example-43).

Consequently, it is reasonable to conclude that having a role of "generic," "none," or "presentation" should not impact the outcome of a test.

Author: @giacomo-petri Can you get a look at it now? I tried to improve the situation…

---
id: j7zzqr
name: ARIA role is permitted for the element.
rule_type: atomic
description: |
  This rule checks that explicit WAI-ARIA roles are allowed for the element they are specified on.
accessibility_requirements:
  html-aria:docconformance:

>     title: ARIA in HTML, 4. Document conformance requirements for use of ARIA attributes in HTML

Reviewer: Sorry, this is going to be very nit picky.

> The third cell in each row defines the ARIA role values and aria-* attributes which authors MAY specify on the element

I don't think the way you're interpreting this requirement is right. Or quite possibly, I think the requirement is written incorrectly. MAY in RFC2119 speak does not create a prohibitive statement. You can use these roles on those elements, and browsers have to support it when you do. But that's not the same as saying authors are not allowed to use other roles. It would have to be something like "if an author specifies a role it MUST be one from the third column."

I think this is a solid rule, and we should definitely have it, but I'm a little reluctant to say this is a conformance requirement for ARIA in HTML. Even if that was the intended meaning, that's not what it says. I don't really want to hold this rule up over a technicality like this, but I do think something needs to be done about this.

Reviewer 4: @WilcoFiers the sentence you quote needs to be put into the context of section 1 Author requirements fo ruse of ARIA in HTML

>Authors MUST NOT use the ARIA role and aria-* attributes in a manner that conflicts with the semantics described in the [4. Document conformance requirements for use of ARIA attributes in HTML](https://w3c.github.io/html-aria/#docconformance) and [4.2 Requirements for use of ARIA attributes in place of equivalent HTML attributes](https://w3c.github.io/html-aria/#docconformance-attr) tables. ...

per section 1, it's an Authors MUST NOT - where the Author MAYs specificy what it allowed.

if you think this is unclear because one has to piece together these different parts of the specification, then that's fair.  The spec can be updated to restate the author requirements more overtly in tandem with the text you quoted.

    forConformance: true
    failed: not satisfied
    passed: satisfied
    inapplicable: satisfied
input_aspects:
  - Accessibility Tree
  - DOM Tree
acknowledgments:
  authors:
    - Jean-Yves Moyen
---

## Applicability

This rule applies to any [HTML element][namespaced element] which has an [explicit semantic role][explicit role] and for which at least one of the following is true:

- the element is [included in the accessibility tree][]; or

> - the element has an [explicit semantic role][] of `presentation` or `none`.

Reviewer: It's a little odd to me that this rule would apply to an element with role=none that's inside an aria-hidden=true container. Perhaps the applicability should just have an exception for elements that are programmatically hidden?

This rule does not check specifically whether the [explicit role][] is deprecated, or whether it is the special `generic` role that should not be used by authors. These are not conformance requirement, or are checked by other rules.

The roles of `none` or `presentation` are only allowed on certain elements. Therefore this rule also check these elements, even though they are not necessarily [included in the accessibility tree][]. Many of the elements for which `presentation` is not allowed are interactive elements where the role would anyway be ignored due to the [presentational roles conflict resolution][].

### Related rules

- [Role attribute has valid value](https://www.w3.org/WAI/standards-guidelines/act/rules/674b10/proposed/) checks that the value of the `role` attribute exists in ARIA, while this rule checks that it is allowed on the element using it.

### Bibliography

- [Document conformance requirements for use of ARIA attributes in HTML](https://www.w3.org/TR/html-aria/#docconformance)

## Test Cases

### Passed

#### Passed Example 1

> This `a` element with an `href` attribute has an [explicit role][] of `button`, which is allowed.

Reviewer: I think these descriptions would read better if we flip the role and element around:

```suggestion
The `button` role is allowed on an `a` element with an `href` attribute.
```

```html
<a href="https://www.w3.org/WAI/standards-guidelines/act/rules/" role="button">All ACT rules</a>
```

#### Passed Example 2

This `h1` element has an [explicit role][] of `tab`, which is allowed

```html

> <h1 role="tab">ACT rules</h1>

Reviewer: I'm not a huge fan of doing this, as there are more requirements for tab to be valid, like that it needs to be in a tablist. I think either a different role is needed here, or this needs to be made a fully valid tab.

Fruits:
<ul>
	<li>
		Apple
		<hr role="presentation" />
	</li>
	<li>
		Banana
		<hr role="presentation" />
	</li>
	<li>Orange</li>
</ul>
```

#### Passed Example 4

This `p` element has an [explicit role][] of `generic` which is allowed. Using an explicit role of `generic` for any element is globally not recommended, but not an error by itself.

```html
<p role="generic">It was a dark and stormy night.</p>
```

#### Passed Example 5

These `hr` elements have an [explicit role][] of `separator`, which is allowed. Using an explicit role which is the same as the [implicit role][] is not recommended, but not an error by itself.

```html
Fruits:
<ul>
	<li>
		Apple
		<hr role="separator" />
	</li>
	<li>
		Banana
		<hr role="separator" />
	</li>
	<li>Orange</li>
</ul>
```

#### Passed Example 6

This `h1` element has an [explicit role][] of `tab`, which is allowed. The `btn` token is not a valid role and is therefore ignored.

```html
<h1 role="btn tab">ACT rules</h1>
```

#### Passed Example 7

This `dialog` element has an [explicit role][] of `alertdialog`, which is allowed. Even though the `alert` role is not allowed, the first token is a valid role and is therefore used.

```html

> <dialog role="alertdialog alert">This is not right.</dialog>

Reviewer: I'm not 100% sure about this one. The requirement for it is this:

> The third cell in each row defines the ARIA role values and aria-* attributes which authors MAY specify on the element.

That to me reads that fallback roles apply here too. I don't know that I would want them too. That feels a little on the strict side, but it might be good to put something in the background section about fallback roles, and straight up invalid or abstract roles being out of scope for this rule, even if they do seem like they'd fail the requirement.

	<a href="https://www.w3.org/WAI/standards-guidelines/act/rules/"> ACT rules</a>
</aside>
```

#### Failed Example 3

These `h1` elements have an [explicit role][] of `listitem`, which is not allowed; the `div` element has an [explicit role][] of `list`, which is allowed.

```html
Fruits:
<div role="list">
	<h1 role="listitem">Apple</h1>
	<h1 role="listitem">Banana</h1>
	<h1 role="listitem">Orange</h1>
</div>
```

#### Failed Example 4

This `a` element with an `href` attribute has an [explicit role][] of `presentation`, which is not allowed. Since the element is focusable, the [presentational roles conflict resolution][] would trigger, and the explicit role is ignored.

```html
<a href="https://www.w3.org/WAI/standards-guidelines/act/rules/" role="presentation">All ACT rules</a>
```

#### Failed Example 5

The first `li` element has an [explicit role][] of `presentation`, which is not allowed when its parent has an [implicit role][] of `list`.

```html
Fruits:
<ul>
	<li role="presentation">Apple</li>
	<li>Banana</li>
	<li>Orange</li>
</ul>
```

#### Failed Example 6

This `label` element has an [explicit role][] of `generic`, which is not allowed.

```html

> <label role="generic"> Name: <input /> </label>

Reviewer: ARIA is a little vague on whether generic elements are supposed to be in the accessibility tree. I think the applicability as you have it now probably means this passes. My suggested update to it should solve that.

Reviewer 4: 'passes' though a should not from aria.

i think aria in html's rules for label are a bit too strict here.  i'm going to make an issue to address this.

Reviewer 3: I understood the point of this example to be that generic is a user-agent assigned role, but not one that you are allowed to explicitly specify. Now that I think about it though, that may be more appropriate for https://www.w3.org/WAI/standards-guidelines/act/rules/674b10/proposed/

Reviewer 4: https://github.com/w3c/html-aria/issues/492

    failed: not satisfied
    passed: satisfied
    inapplicable: satisfied
input_aspects:
  - Accessibility Tree
  - DOM Tree
acknowledgments:
  authors:
    - Jean-Yves Moyen
---

## Applicability

This rule applies to any [HTML element][namespaced element] which has an [explicit semantic role][explicit role] and is not [programmatically hidden][].

## Expectation

For each test target, its [explicit semantic role][explicit role] is allowed on this element, according to the [ARIA in HTML specification][aria in html document conformance].

## Assumptions

There are no assumptions.

## Accessibility Support

There are no accessibility support issues known.

## Background

Each element also has an [implicit semantic role][implicit role] defined in the [HTML Accessibility API Mappings][html aam:roles]. Setting the [explicit role][] as the same as the [implicit role][] is not recommended but nonetheless allowed. This rule does not check that scenario.

This rule does not check specifically whether the [explicit role][] is deprecated, or whether it is the special `generic` role that should not be used by authors. These are not conformance requirement, or are checked by other rules.

> This rule only check for elements that are not [programmatically hidden][]. Elements that are [programmatically hidden][], e.g. due to a CSS `display:none`, must still have valid roles in order to conform to [ARIA in HTML][aria in html document conformance]. However, these won't cause any actual accessibility issues as long as they stay hidden. This rule only checks a snapshot of a page in a given state; if interacting with the page reveals such an element without fixing its role, that could create an accessibility issue for the user. Examples only show [programmatically hidden][] element with a correct role.

Reviewer 4: i appreciate the note being added.  I've made some wording updates that I hope can be merged.  Let me know if there are any questions.

```suggestion
This rule only check for elements that are not [programmatically hidden][]. Elements that are [programmatically hidden][], e.g. due to a CSS `display:none`, must still have valid roles in order to conform to [ARIA in HTML][aria in html document conformance]. However, as long as the elements with the errant roles remain programmatically hidden, they are not expected to cause accessibility issues. This rule only checks a snapshot of a page in a given state; if interacting with the page programmatically exposes an element with an errant role that this rule previously ignored, the new state of the page should be reevaluated to ensure it still passes this rule, and that an accessibility issue has not been introduced to users. Examples only show [programmatically hidden][] elements with a correct role.
```

> This rule only check for elements that are not [programmatically hidden][]. Elements that are [programmatically hidden][], e.g. due to a CSS `display:none`, must still have valid roles in order to conform to [ARIA in HTML][aria in html document conformance]. However, as long as the elements with the errant roles remain programmatically hidden, they are not expected to cause accessibility issues. This rule only checks a snapshot of a page in a given state; if interacting with the page programmatically exposes an element with an errant role that this rule previously ignored, the new state of the page should be reevaluated to ensure it still passes this rule, and that an accessibility issue has not been introduced to users. Examples only show [programmatically hidden][] elements with a correct role.

Reviewer 4: just a missing 's' i had previously not noticed in the first sentence.  otherwise lgtm.

```suggestion
This rule only checks for elements that are not [programmatically hidden][]. Elements that are [programmatically hidden][], e.g. due to a CSS `display:none`, must still have valid roles in order to conform to [ARIA in HTML][aria in html document conformance]. However, as long as the elements with the errant roles remain programmatically hidden, they are not expected to cause accessibility issues. This rule only checks a snapshot of a page in a given state; if interacting with the page programmatically exposes an element with an errant role that this rule previously ignored, the new state of the page should be reevaluated to ensure it still passes this rule, and that an accessibility issue has not been introduced to users. Examples only show [programmatically hidden][] elements with a correct role.
```
